lib/classes/batch/class-batch-task-manager.php

```diff
@@ -0,0 +1,353 @@
+      $description = $object->get_description();
+
+      $result = $object->process_item($item);
+      $this->_update_state( $object->get_state() );
+
+      $result = apply_filters('wp_stateless_batch_task_item_processed', $result, $item);
+    } catch (\Throwable $e) {
+      Helper::log( "Batch task unable to handle item $item: " . $e->getMessage() );
+      
+      $result = apply_filters('wp_stateless_batch_task_item_failed', $result, $item);
+    }
+
+    return $result;
+  }
+  
+  /**
+   * Complete the batch task. Tries to get the next batch and continue
+   */
+  protected function complete() {
+    $class = '';
+    $description = 'Batch process';
+
+    // Check if we have more batched to run
+    try {
+      $object = $this->_get_batch_task_object();
+      $class = $object::class;  
+      $batch = $object->get_batch();
+      $description = $object->get_description();
+
+      if ( !empty($batch) ) {
+        $this->_add_batch( $batch );
+        $this->_update_state( $object->get_state() );
+
+        $this->dispatch();
+
+        return;
+      }
+
+      Helper::log( 'Batch task completed: ' . $object::class );
+    } catch (\Throwable $e) {
+      Helper::log( "Unable to process next batch: " . $e->getMessage() );
+    }
+
+    // If no more batches - delete state
+    $state = $this->_get_state();
+
+    parent::complete();
+    $this->_delete_state();
+
+    do_action('wp_stateless_batch_task_finished', $class);
+
+    $site = site_url();
+    $subject = sprintf( __('Stateless %s complete successfully', ud_get_stateless_media()->domain), $description );
+    $message = sprintf(
+      __("This is a simple notification to inform you that the WP-Stateless plugin has finished a %s process for %s.\n\nIf you have WP_DEBUG_LOG enabled, check those logs to review any errors that may have occurred during the process.", ud_get_stateless_media()->domain), 
+      $description, 
+      $site
+    );
+
+    do_action('wp_stateless_send_admin_email', $subject, $message, $state['email'] ?? '');
+  }
+
+  /**
+   * Check if batch task has a state, so it is in progress
+   * Because is_processing is true only while processing an item
+   * 
+   * @param array|null $state
+   * @return bool
+   */
+  public function is_running($state = null) {
+    if ( empty($state) ) {
+      $state = $this->_get_state();
+    }
+
+    return !empty($state);
+  }
+
+  /**
+   * Get the state of the current batch process
+   * 
+   * @param mixed $status
+   * @return mixed
+   */
+  public function get_state($state) {
+    $state = $this->_get_state();
+
+    unset($state['class']);
+    unset($state['file']);
+
+    $state['is_running'] = $this->is_running($state);
+    $state['is_paused'] = $this->is_paused();
+
+    return $state;
+  }
+
+  /**
+   * Pause the batch task
+   * 
+   * @param array $state
+   * @param array $params
+   * @return array
+   */
+  public function pause_task($state, $params) {
```

Comment:
@balexey88 looks like `$params` can be removed here

Reply:
This is for future use. There is one API interface for all actions like start, pause, resume, etc. And currently batch task manager works with only one task at a time.
In the future we can extend it allowing to run several tasks at once, add cancel functionality and so on.
So this additional `$params` will be used in the future, passing some data between tasks and task manager using hooks chain.